cli/src/commands/dsl/validate.ts

```diff
@@ -153,7 +160,9 @@ export default class Validate extends BaseCommand {
         ${fileNameList.length} 
         file${fileNameList.length > 1 ? "s" : ""}, exiting program`
       );
-      this.exit(1);
+      if (severity === "error") {
```

Comment:
Does it make more sense to check all the errors and warnings and only do `this.exit(1)` if any errors exist, allowing for warnings? This'd let the validator dynamically pass/fail depending on if there are actually errors vs warnings.

Reply:
second approach:
```ts
/** A command thay runs TS typechecker against source ts and tsx files */
export default class Validate extends BaseCommand {
  static description = "Validate TSX files before they get compiled";

  static flags = {
    ...BaseCommand.flags,
    files: Flags.string({
      char: "f",
      description: "A list of files or globs to validate",
      multiple: true,
    }),
    severity: Flags.string({
      char: "s",
      description: "The severity of the validation",
      options: ["error", "warn"],
      default: "error",
    }),
  };

  private async getOptions() {
    const { flags } = await this.parse(Validate);
    const config = await this.getPlayerConfig();

    const files =
      flags.files && flags.files.length > 0 ? flags.files : config.dsl?.src;

    if (!files) {
      throw new Error("DSL TSC typechecking requires a file list");
    }

    return {
      inputFiles: Array.isArray(files) ? files : [files],
      severity: flags.severity,
    };
  }

  private getTSConfig(filePath?: string): ts.CompilerOptions | undefined {
    const configFileLocation = ts.findConfigFile(
      filePath ?? ".",
      (fName: string) => {
        return existsSync(fName);
      }
    );

    if (!configFileLocation) {
      this.debug(`No TSConfig found`);
      return;
    }

    const configContent = ts.readConfigFile(configFileLocation, (p) => {
      return readFileSync(p, "utf-8");
    });

    if (configContent.error) {
      this.warn(
        ts.flattenDiagnosticMessageText(configContent.error.messageText, "\n")
      );

      return;
    }

    const basePath = path.dirname(configFileLocation);
    const parsedConfigContent = ts.parseJsonConfigFileContent(
      configContent.config,
      ts.sys,
      basePath
    );

    if (parsedConfigContent.errors.length > 0) {
      throw new Error(
        `Error while parsing tsconfig: ${parsedConfigContent.errors
          .map((d) => {
            return ts.flattenDiagnosticMessageText(d.messageText, "\n");
          })
          .join("\n\n")}`
      );
    }

    return parsedConfigContent.options;
  }

  async run(): Promise<void> {
    const { inputFiles, severity } = await this.getOptions();

    const files = await glob(
      convertToFileGlob(inputFiles, "**/*.(tsx|jsx|js|ts)"),
      {
        expandDirectories: true,
      }
    );

    const TSConfig = this.getTSConfig() ?? DEFAULT_COMPILER_OPTIONS;

    const program = ts.createProgram(files, TSConfig);

    const allDiagnostics = ts.getPreEmitDiagnostics(program);

    let diagnosticsCount = 0;

    const groupedDiagnostics = allDiagnostics.reduce(
      (acc, diagnostic) => {
        const fileName = diagnostic.file?.fileName;
        const category = diagnostic.category;

        if (fileName && files.includes(fileName)) {
          const key =
            category === ts.DiagnosticCategory.Error ? "errors" : "warnings";

          if (!acc[key][fileName]) {
            acc[key][fileName] = [];
          }

          acc[key][fileName].push(diagnostic);
          diagnosticsCount += 1;
        }

        return acc;
      },
      { errors: {}, warnings: {} } as {
        errors: Record<string, ts.Diagnostic[]>;
        warnings: Record<string, ts.Diagnostic[]>;
      }
    );

    ["errors", "warnings"].forEach((category) => {
      const diagnosticsList = Object.keys(
        groupedDiagnostics[category as keyof typeof groupedDiagnostics]
      );

      diagnosticsList.forEach((diagnosticGroup) => {
        this.log(`${diagnosticGroup}`);
        groupedDiagnostics[category as keyof typeof groupedDiagnostics][
          diagnosticGroup
        ].forEach((diagnostic) => {
          if (diagnostic.file) {
            const { line, character } = ts.getLineAndCharacterOfPosition(
              diagnostic.file,
              diagnostic.start ?? 0
            );
            const message = ts.flattenDiagnosticMessageText(
              diagnostic.messageText,
              "\n"
            );

            const logSymbol =
              category === "errors" ? logSymbols.error : logSymbols.warning;

            this.log(
              `  ${logSymbol} (${line + 1},${character + 1}): ${message}`
            );
          } else {
            this.log(
              ts.flattenDiagnosticMessageText(diagnostic.messageText, "\n")
            );
          }
        });
      });
    });

    const errorsCount = Object.keys(groupedDiagnostics.errors).length;
    const warningsCount = Object.keys(groupedDiagnostics.warnings).length;

    if (errorsCount) {
      this.log(
        `Type or syntax errors found in ${errorsCount} file${
          errorsCount > 1 ? "s" : ""
        }, exiting program`
      );
      this.exit(1);
    } else if (warningsCount) {
      this.log(
        `Type or syntax warnings found in ${warningsCount} file${
          warningsCount > 1 ? "s" : ""
        }, ${severity === "warn" ? "exiting program" : "continuing"}`
      );
      if (severity === "warn") {
        this.exit(1);
      }
    } else {
      this.log(`${logSymbols.success} No TSX types or errors found.`);
    }
  }
}
```

Reply:
>Although we could use the diagnostic.category enumeration where Error = 1, I suggest using severity to define the behaviour during the validation step:
>
>If the severity is set to "warn," we will only log the diagnostic.
>If it's set to "error" (default), the process will exit with code 1.

This would be my expected behavior as well.